### PR TITLE
Forward Error Correction

### PR DESCRIPTION
**Pull request by irtimmer**
This patch set contains code changes needed to support FEC. The code to do the actual error correction is still missing as I currently only get it to work with a single FEC packet per frame.

A new queue is added which buffers the whole frame as error correction is done on whole frames. Reordening of packets therefore only works for packets in the same frame and instead of keeping a max size and max time the receiving of the next frame is used to flush the packets from the current frame to the video depacketizer.

### CODE REVIEW

**Comment by irtimmer** [comment]
Initial implementation for doing error-correction with single FEC packet has been added

**Comment by irtimmer** [comment]
The problems with multiple FEC packets have been solved. So most of the code is ready, although their are probably still some places on where the FEC implementation could use some improvements

File: src/RtpFecQueue.c [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+    int totalPackets = ushort(queue->bufferHighestSequenceNumber - queue->bufferLowestSequenceNumber) + 1;
+    int totalParityPackets = (queue->bufferDataPackets * queue->fecPercentage + 99) / 100;
+    int parityPackets = totalPackets - queue->bufferDataPackets;
+    int missingPackets = totalPackets - queue->bufferSize;
+    
+    if (parityPackets < missingPackets || missingPackets <= 0) {
+        return;
+    }
+    
+    reed_solomon* rs = reed_solomon_new(queue->bufferDataPackets, totalParityPackets);
```
Comment (cgutman) [inline src/RtpFecQueue.c]: Check for NULL

File: src/RtpFecQueue.c [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    rs->shards = queue->bufferDataPackets + missingPackets; //Don't let RS complain about missing parity packets
+
+    int* missing = malloc(missingPackets * sizeof(int));
+    unsigned char** packets = malloc(totalPackets * sizeof(unsigned char**));
+    unsigned char* marks = malloc(totalPackets * sizeof(unsigned char**));
```
Comment (cgutman) [inline src/RtpFecQueue.c]: Should be `sizeof(unsigned char)`

File: src/RtpFecQueue.c [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+            memset(&packets[index][size], 0, receiveSize - size);
+        }
+
+        entry = entry->next;
+    }
+
+    int i;
+    for (i = 0; i < totalPackets; i++) {
+        if (marks[i]) {
+            packets[i] = malloc(bufferSize);
```
Comment (cgutman) [inline src/RtpFecQueue.c]: Check for NULL

File: src/RtpFecQueue.c [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+    int* missing = malloc(missingPackets * sizeof(int));
+    unsigned char** packets = malloc(totalPackets * sizeof(unsigned char**));
+    unsigned char* marks = malloc(totalPackets * sizeof(unsigned char**));
+    if (missing == NULL || packets == NULL || marks == NULL)
+        goto cleanup;
+
+    memset(marks, 1, sizeof(char) * (totalPackets));
+    
+    int receiveSize = StreamConfig.packetSize + MAX_RTP_HEADER_SIZE;
+    int bufferSize = bufferSize = receiveSize + sizeof(int) + sizeof(RTPFEC_QUEUE_ENTRY);
```
Comment (cgutman) [inline src/RtpFecQueue.c]: Extra `bufferSize`

File: src/RtpFecQueue.c [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        if (marks[i]) {
+            packets[i] = malloc(bufferSize);
+        }
+    }
+    
+    int ret = reed_solomon_reconstruct(rs, packets, marks, totalPackets, receiveSize);
+    
+    for (i = 0; i < totalPackets; i++) {
+        if (marks[i]) {
+            if (ret == 0) {
```
Comment (cgutman) [inline src/RtpFecQueue.c]: Should we even enter this loop if `ret != 0`?

File: src/RtpFecQueue.c [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+            queue->queueTail = queue->bufferTail;
+        }
+        queue->bufferHead = NULL;
+        queue->bufferTail = NULL;
+
+        queue->queueSize = queue->bufferSize;
+        queue->nextRtpSequenceNumber = queue->bufferHighestSequenceNumber;
+        
+        int fecIndex = (nvPacket->fecInfo & 0xFF000) >> 12;
+        queue->bufferLowestSequenceNumber = packet->sequenceNumber - fecIndex;
```
Comment (cgutman) [inline src/RtpFecQueue.c]: What happens if the first packet in a frame is dropped? Would this make `queue->bufferLowestSequenceNumber` wrong?

File: src/RtpFecQueue.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
+#pragma once
+
+#include "Video.h"
+
+typedef struct _RTPFEC_QUEUE_ENTRY {
+    PRTP_PACKET packet;
+
+    uint64_t queueTimeMs;
```
Comment (cgutman) [inline src/RtpFecQueue.h]: I don't really think it makes much sense to track this in the FEC queue. We'll automatically move on to the next frame when it arrives.

File: src/RtpFecQueue.c [hunk 8]
```diff
@@ -0,0 +1,291 @@
+#include "Limelight-internal.h"
+#include "RtpFecQueue.h"
+#include "rs.h"
+
+#define ushort(x) (unsigned short) (x % (UINT16_MAX+1))
+#define isBefore(x, y) ushort(x - y) > (UINT16_MAX/2)
```
Comment (cgutman) [inline src/RtpFecQueue.c]: Also enclose the whole operation itself in parenthesis too.

File: reedsolomon/rs.c [hunk 9]
```diff
@@ -0,0 +1,979 @@
...
+    for (row = 0 ; row < k ; row++) {
+        if (ipiv[row] != 1) {
+        for (ix = 0 ; ix < k ; ix++) {
+            DEB( pivloops++ ; )
+            if (ipiv[ix] == 0) {
+                if (src[row*k + ix] != 0) {
+                irow = row ;
+                icol = ix ;
+                goto found_piv ;
+                }
```
Comment (cgutman) [inline reedsolomon/rs.c]: Looks like this function has some indentation issues.

File: reedsolomon/rs.c [hunk 10]
```diff
@@ -0,0 +1,979 @@
...
+            p[icol] = 0 ;
+            addmul(p, pivot_row, c, k );
+        }
+        }
+    }
+    id_row[icol] = 0;
+    } /* done all columns */
+    for (col = k-1 ; col >= 0 ; col-- ) {
+    if (indxr[col] <0 || indxr[col] >= k)
+        fprintf(stderr, "AARGH, indxr[col] %d\n", indxr[col]);
```
Comment (cgutman) [inline reedsolomon/rs.c]: We shouldn't be printing here.

File: src/RtpFecQueue.c [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+       dataOffset += 4; // 2 additional fields
+    }
+
+    PNV_VIDEO_PACKET nvPacket = (PNV_VIDEO_PACKET)(((char*)packet) + dataOffset);
+    nvPacket->frameIndex = queue->currentFrameNumber;
+    
+    //Set size of generated packet
+    int size = StreamConfig.packetSize + dataOffset;
+    
+    //Remove padding if it's the last packet in frame
```
Comment (irtimmer) [inline src/RtpFecQueue.c]: No, I already encountered some artifacts myself. So probably I just remove this.

File: src/RtpFecQueue.c (hunk 5 above)
Comment (irtimmer) [inline src/RtpFecQueue.c]: Yes, because in that case we use it to cleanup the malloced packets

File: src/RtpFecQueue.c (hunk 6 above)
Comment (irtimmer) [inline src/RtpFecQueue.c]: No, therefore we substract the fecIndex

**Comment by irtimmer** [comment]
Thank you for reviewing the code. The reedsolomon code has been copied from https://github.com/jannson/reedsolomon-c including the fprintf and indention issues ;). Minimal changes has been made to make it work with GameStream. Soon I hope to cleanup the code a little bit more.

File: src/RtpFecQueue.c [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+                goto cleanup;
+            }
+        }
+    }
+    
+    ret = reed_solomon_reconstruct(rs, packets, marks, totalPackets, receiveSize);
+
+    cleanup:
+    for (i = 0; i < totalPackets; i++) {
+        if (marks[i]) {
```
Comment (d3m3vilurr) [inline src/RtpFecQueue.c]: `marks` can be null.

File: src/RtpFecQueue.c [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+                if (rtpPacket->sequenceNumber == ushort(queue->bufferLowestSequenceNumber + queue->bufferDataPackets - 1)) {
+                    while (packets[i][size-1] == 0) {
+                        size--;
+                    }
+                }
+
+                memcpy(&packets[i][receiveSize], &size, sizeof(int));
+                queuePacket(queue, queueEntry, 0, rtpPacket);
+            } else {
+                free(packets[i]);
```
Comment (d3m3vilurr) [inline src/RtpFecQueue.c]: `packets` also can be null.

**Pull request merged**